### PR TITLE
feat(model): refine paddle code and support auxiliary_vars in model.py

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
from PR https://github.com/lululxvi/deepxde/pull/1070

1. support auxiliary_vars and refine paddle code in model.py

### CODE REVIEW

File: deepxde/model.py [hunk 1]
```diff
@@ -260,10 +265,10 @@ def outputs(training, inputs):
             with torch.no_grad():
                 if isinstance(inputs, tuple):
                     inputs = tuple(
-                        map(lambda x: torch.as_tensor(x).requires_grad_(), inputs)
+                        map(lambda x: torch.as_tensor(x, dtype=torch.float32).requires_grad_(), inputs)
```
Comment (lululxvi) [inline deepxde/model.py:268]: Don't hard code float32. DeepXDE supports float32 and float64 via `dde.config.set_default_float`

Comment (HydrogenSulfate) [reply deepxde/model.py]: Done

File: deepxde/model.py [hunk 2]
```diff
@@ -260,10 +266,10 @@ def outputs(training, inputs):
             with torch.no_grad():
                 if isinstance(inputs, tuple):
                     inputs = tuple(
-                        map(lambda x: torch.as_tensor(x).requires_grad_(), inputs)
+                        map(lambda x: torch.as_tensor(x, dtype=config.real(bkd.lib)).requires_grad_(), inputs)
```
[the remark this answers is not in the document]
Comment (HydrogenSulfate) [reply deepxde/model.py]: Done

File: deepxde/config.py [hunk 3]
```diff
@@ -47,6 +47,10 @@ def set_default_float(value):
         real.set_float64()
     if backend_name in ["tensorflow.compat.v1", "tensorflow"]:
         tf.keras.backend.set_floatx(value)
+    elif backend_name in ["paddle"]:
```
Comment (lululxvi) [inline deepxde/config.py:50]: `backend_name == "paddle"`

File: deepxde/config.py [hunk 4]
```diff
@@ -47,6 +47,10 @@ def set_default_float(value):
...
+        paddle.set_default_dtype(value)
+    elif backend_name in ["pytorch"]:
```
Comment (lululxvi) [inline deepxde/config.py:52]: `backend_name = "pytorch"`

File: deepxde/config.py [hunk 5]
```diff
@@ -47,6 +47,10 @@ def set_default_float(value):
...
+        torch.set_default_dtype(real(bkd.lib))
```
Comment (lululxvi) [inline deepxde/config.py:53]: `torch.set_default_dtype(real(torch))`

Comment (lululxvi) [inline deepxde/config.py:53]: Swap the order of pytorch and paddle

Comment (HydrogenSulfate) [reply deepxde/config.py]: Done

Comment (HydrogenSulfate) [reply deepxde/config.py]: Done

File: deepxde/data/pde.py [hunk 6]
```diff
@@ -160,10 +160,10 @@ def losses(self, targets, outputs, loss_fn, inputs, model, aux=None):
     @run_if_all_none("train_x", "train_y", "train_aux_vars")
     def train_next_batch(self, batch_size=None):
         self.train_x_all = self.train_points()
-        self.train_x = self.bc_points()
+        self.train_x = self.bc_points().astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/data/pde.py:163]: `.astype(config.real(np))` is not necessary. `self.bc_points()` should return the correctly dtype. The same for other cases.

If there is some code raising an error about dtype, then we need to track back to the first place where dtype gets wrong and fix it.

Comment (HydrogenSulfate) [reply deepxde/data/pde.py]: Done

File: deepxde/data/ide.py [hunk 7]
```diff
@@ -81,8 +81,8 @@ def train_next_batch(self, batch_size=None):
         self.train_x_all = self.train_points()
         x_bc = self.bc_points()
         x_quad = self.quad_points(self.train_x_all)
-        self.train_x = np.vstack((x_bc, self.train_x_all, x_quad))
-        self.train_y = self.soln(self.train_x) if self.soln else None
+        self.train_x = np.vstack((x_bc, self.train_x_all, x_quad)).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/data/ide.py:84]: `.astype(config.real(np))` is not necessary. If the dtype is not correct, then it should be due to `x_quad`, and you can add type conversion in `self.quad_points()`.

File: deepxde/data/pde.py [hunk 8]
```diff
@@ -194,7 +191,7 @@ def resample_train_points(self, pde_points=True, bc_points=True):
 
     def add_anchors(self, anchors):
         """Add new points for training PDE losses. The BC points will not be updated."""
-        anchors = anchors.astype(config.real(np))
+        anchors = anchors
```
Comment (lululxvi) [inline deepxde/data/pde.py:194]: Type conversion is necessay, as `anchors` are provided by user.

File: deepxde/utils/array_ops_compat.py [hunk 9]
```diff
@@ -27,7 +27,7 @@ def hstack(tup):
             tup[0] = tf.convert_to_tensor([], dtype=config.real(tf))
         else:
             tup[0] = np.array([], dtype=config.real(np))
-    return tf.concat(tup, 0) if is_tensor(tup[0]) else np.hstack(tup)
+    return tf.concat(tup, 0) if is_tensor(tup[0]) else np.hstack(tup).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/utils/array_ops_compat.py:30]: `.astype(config.real(np))` doesn't seem necessary.

File: deepxde/data/ide.py (hunk 7 above)
Comment (HydrogenSulfate) [reply deepxde/data/ide.py]: Done

File: deepxde/data/ide.py [hunk 10]
```diff
@@ -92,8 +92,8 @@ def test(self):
         else:
             self.test_x = self.test_points()
         x_quad = self.quad_points(self.test_x)
-        self.test_x = np.vstack((self.test_x, x_quad))
-        self.test_y = self.soln(self.test_x) if self.soln else None
+        self.test_x = np.vstack((self.test_x, x_quad)).astype(config.real(np))
```
[the remark this answers is not in the document]
Comment (HydrogenSulfate) [reply deepxde/data/ide.py]: Done

File: deepxde/data/pde.py [hunk 11]
```diff
@@ -165,9 +165,7 @@ def train_next_batch(self, batch_size=None):
             self.train_x = np.vstack((self.train_x, self.train_x_all))
         self.train_y = self.soln(self.train_x) if self.soln else None
         if self.auxiliary_var_fn is not None:
-            self.train_aux_vars = self.auxiliary_var_fn(self.train_x).astype(
-                config.real(np)
-            )
+            self.train_aux_vars = self.auxiliary_var_fn(self.train_x)
```
[the remark this answers is not in the document]
Comment (HydrogenSulfate) [reply deepxde/data/pde.py]: restored

File: deepxde/data/pde.py (hunk 8 above)
Comment (HydrogenSulfate) [reply deepxde/data/pde.py]: restored

File: deepxde/utils/array_ops_compat.py (hunk 9 above)
Comment (HydrogenSulfate) [reply deepxde/utils/array_ops_compat.py]: Done

File: deepxde/optimizers/paddle/optimizers.py [hunk 12]
```diff
@@ -12,18 +38,16 @@ def get(params, optimizer, learning_rate=None, decay=None):
     if isinstance(optimizer, paddle.optimizer.Optimizer):
         return optimizer
 
-    if optimizer in ["L-BFGS", "L-BFGS-B"]:
-        # TODO: add support for L-BFGS and L-BFGS-B
-        raise ValueError("L-BFGS to be implemented for backend Paddle.")
+    if is_external_optimizer(optimizer):
```
Comment (lululxvi) [inline deepxde/optimizers/paddle/optimizers.py:41]: Keep the TODO comment

File: deepxde/optimizers/paddle/optimizers.py [hunk 13]
```diff
@@ -3,6 +3,32 @@
 import paddle
 
 
+class InverseTimeDecay(paddle.optimizer.lr.InverseTimeDecay):
+    def __init__(self, learning_rate, gamma, decay_steps=1, last_epoch=-1, verbose=False):
+        self.decay_steps = decay_steps
+        super().__init__(learning_rate, gamma, last_epoch, verbose)
+
+    def get_lr(self):
```
Comment (lululxvi) [inline deepxde/optimizers/paddle/optimizers.py:11]: Why implementing `get_lr`? Is it in `paddle.optimizer.lr.InverseTimeDecay`?

Comment (HydrogenSulfate) [reply deepxde/optimizers/paddle/optimizers.py:11]: > paddle.optimizer.lr.InverseTimeDecay

Actually `paddle.optimizer.lr.InverseTimeDecay` is slightly different from `tf.train.inverse_time_decay`, as below.


```
paddle 2.x : lr=initial_learning_rate/(1+decay_rate∗epoch)
tf/padle1.x: lr=initial_learning_rate/(1+decay_rate*(step/decay_step))
```

[tensorflow InverseTimeDecay](https://tensorflow.google.cn/api_docs/python/tf/keras/optimizers/schedules/InverseTimeDecay?hl=en)
[paddle 1.8 InverseTimeDecay](https://www.paddlepaddle.org.cn/documentation/docs/zh/1.8/api_cn/dygraph_cn/InverseTimeDecay_cn.html#inversetimedecay)
[paddle 2.4 InverseTimeDecay](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/optimizer/lr/InverseTimeDecay_cn.html#inversetimedecay)

so I derive `paddle.optimizer.lr.InverseTimeDecay` and reimplement `get_lr()` for omitted argument `decay_step`, so cound be aligned with `tf.train.inverse_time_decay`

Comment (lululxvi) [reply deepxde/optimizers/paddle/optimizers.py:11]: We don't have to align paddle with tf. The learning rate in different backends are very different, so it is impossible to align them. We have the docs of lr decay for each backend in Model.compile that users can check the usage.

File: deepxde/data/fpde.py [hunk 14]
```diff
@@ -448,7 +448,7 @@ def get_x_dynamic(self):
             # xi, wi = zip(*map(self.modify_second_order, xi, wi))
             # third order
             # xi, wi = zip(*map(self.modify_third_order, xi, wi))
-            x.append(np.vstack(xi))
+            x.append(np.vstack(xi).astype(config.real(np)))
```
Comment (lululxvi) [inline deepxde/data/fpde.py:451]: If the dtype is not corrent, then it seems that it is from line 432, i.e., we should convert the dtype in geom.background_points

File: deepxde/model.py [hunk 15]
```diff
@@ -4,7 +4,6 @@
 from collections import OrderedDict
 
 import numpy as np
-
 from . import config
```
Comment (lululxvi) [inline deepxde/model.py:7]: Keep the empty line 7.

File: deepxde/model.py [hunk 16]
```diff
@@ -53,6 +51,9 @@ def __init__(self, data, net):
         elif backend_name == "jax":
             self.opt_state = None
             self.params = None
+        elif backend_name == "paddle":
```
Comment (lululxvi) [inline deepxde/model.py:54]: This can be combined with pytorch line 49.

File: deepxde/model.py [hunk 17]
```diff
@@ -619,6 +637,7 @@ def _train_sgd(self, iterations, display_every):
             self.train_state.epoch += 1
             self.train_state.step += 1
             if self.train_state.step % display_every == 0 or i + 1 == iterations:
+                self.train_state.set_data_test(*self.data.test())
```
Comment (lululxvi) [inline deepxde/model.py:640]: Not necessay. see line 599 in Model.train

File: deepxde/data/fpde.py (hunk 14 above)
Comment (lululxvi) [reply deepxde/data/fpde.py:453]: If dtype has been fixed, then `.astype(config.real(np)` can be deleted

File: deepxde/optimizers/paddle/optimizers.py [hunk 18]
```diff
@@ -3,6 +3,22 @@
 import paddle
 
 
+def _get_lr_scheduler(lr, decay):
+    if decay is None:
+        return lr, None
```
Comment (lululxvi) [inline deepxde/optimizers/paddle/optimizers.py:8]: Lines 7 and 8 are not necessary.

File: deepxde/geometry/geometry_2d.py [hunk 19]
```diff
@@ -62,10 +62,10 @@ def random_boundary_points(self, n, random="pseudo"):
 
     def background_points(self, x, dirn, dist2npt, shift):
         dirn = dirn / np.linalg.norm(dirn)
-        dx = self.distance2boundary_unitdirn(x, -dirn)
+        dx = self.distance2boundary_unitdirn(x, -dirn).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:65]: `.astype(config.real(np))` should be added in `self.distance2boundary_unitdirn`

File: deepxde/geometry/geometry_nd.py [hunk 20]
```diff
@@ -154,14 +154,14 @@ def random_boundary_points(self, n, random="pseudo"):
...
             U = sample(n, self.dim, random)  # Error for [0, 0, ...] or [0.5, 0.5, ...]
-            X = stats.norm.ppf(U)
+            X = stats.norm.ppf(U).astype(config.real(np))
         X = preprocessing.normalize(X)
         return self.radius * X + self.center
 
     def background_points(self, x, dirn, dist2npt, shift):
         dirn = dirn / np.linalg.norm(dirn)
-        dx = self.distance2boundary_unitdirn(x, -dirn)
+        dx = self.distance2boundary_unitdirn(x, -dirn).astype(config.real(np))
```
Comment (lululxvi) [inline deepxde/geometry/geometry_nd.py:163]: `.astype(config.real(np))` should be added in `self.distance2boundary_unitdirn`

File: deepxde/model.py [hunk 21]
```diff
@@ -94,6 +94,12 @@ def compile(
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
 
+                - For backend PaddlePaddle:
+
+                    - `InverseTimeDecay
+                      <https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/optimizer/lr/InverseTimeDecay_cn.html#inversetimedecay>`_:
+                      ("inverse time", decay_steps, decay_rate)
```
Comment (lululxvi) [inline deepxde/model.py:101]: There is only ` gamma` at https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/optimizer/lr/InverseTimeDecay_en.html

File: deepxde/model.py [hunk 22]
```diff
@@ -94,6 +94,12 @@ def compile(
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
 
+                - For backend PaddlePaddle:
+
+                    - `InverseTimeDecay
+                      <https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/optimizer/lr/InverseTimeDecay_cn.html#inversetimedecay>`_:
+                      ("inverse time", decay_steps, gamma)
```
Comment (lululxvi) [inline deepxde/model.py:101]: Is `decay_steps` needed?

File: deepxde/data/fpde.py (hunk 14 above)
Comment (lululxvi) [reply deepxde/data/fpde.py:453]: I see. But here `xi` is a numpy array, not a scaler?

File: deepxde/model.py (hunk 22 above)
Comment (lululxvi) [reply deepxde/model.py:101]: I think we can just follow paddle's API without using `decay_steps`. Users should check the paddle website for the lr scheduler.

File: deepxde/data/fpde.py (hunk 14 above)
Comment (HydrogenSulfate) [reply deepxde/data/fpde.py]: > I see. But here `xi` is a numpy array, not a scaler?

Oh this line is ommited before, now deleted.

File: deepxde/model.py (hunk 22 above)
Comment (HydrogenSulfate) [reply deepxde/model.py]: > I think we can just follow paddle's API without using `decay_steps`. Users should check the paddle website for the lr scheduler.

Done

File: deepxde/optimizers/paddle/optimizers.py [hunk 23]
```diff
@@ -3,6 +3,20 @@
 import paddle
 
 
+def _get_lr_scheduler(lr, decay):
+    if decay[0] == "inverse time":
+        lr_sch = paddle.optimizer.lr.InverseTimeDecay(
+            lr,
```
Comment (lululxvi) [inline deepxde/optimizers/paddle/optimizers.py:9]: Format via black

File: deepxde/model.py [hunk 24]
```diff
@@ -94,6 +94,12 @@ def compile(
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
 
+                - For backend PaddlePaddle:
+
+                    - `InverseTimeDecay
+                      <https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/optimizer/lr/InverseTimeDecay_cn.html#inversetimedecay>`_:
```
Comment (lululxvi) [inline deepxde/model.py:100]: Use the English doc link

File: deepxde/model.py [hunk 25]
```diff
@@ -508,9 +521,11 @@ def _train_step(self, inputs, targets, auxiliary_vars):
             self.sess.run(self.train_step, feed_dict=feed_dict)
         elif backend_name == "tensorflow":
             self.train_step(inputs, targets, auxiliary_vars)
-        elif backend_name in ["pytorch", "paddle"]:
+        elif backend_name == "pytorch":
             # TODO: auxiliary_vars
             self.train_step(inputs, targets)
+        elif backend_name == "paddle":
+            self.train_step(inputs, targets, auxiliary_vars)
         elif backend_name == "jax":
```
Comment (lululxvi) [inline deepxde/model.py:529]: swap the order of jax and paddle

Comment (HydrogenSulfate) [reply deepxde/model.py:527]: Done

File: deepxde/model.py (hunk 24 above)
Comment (HydrogenSulfate) [reply deepxde/model.py]: Done

File: deepxde/optimizers/paddle/optimizers.py (hunk 23 above)
Comment (HydrogenSulfate) [reply deepxde/optimizers/paddle/optimizers.py]: Done

File: deepxde/model.py [hunk 26]
```diff
@@ -16,7 +16,6 @@
 from .callbacks import CallbackList
 from .utils import list_to_str
 
-
```
Comment (lululxvi) [inline deepxde/model.py:19]: Add this empty line back

Comment (HydrogenSulfate) [reply deepxde/model.py]: Done

File: deepxde/model.py [hunk 27]
```diff
@@ -48,12 +48,13 @@ def __init__(self, data, net):
...
             self.sess = None
             self.saver = None
-        elif backend_name == "pytorch":
+        elif backend_name in ["pytorch", "paddle"]:
             self.lr_scheduler = None
         elif backend_name == "jax":
             self.opt_state = None
             self.params = None
 
+
```
Comment (lululxvi) [inline deepxde/model.py:57]: Delete this emply line.

File: deepxde/model.py [hunk 28]
```diff
@@ -517,6 +531,8 @@ def _train_step(self, inputs, targets, auxiliary_vars):
                 self.params, self.opt_state, inputs, targets
             )
             self.net.params, self.external_trainable_variables = self.params
+        elif backend_name == "paddle":
```
Comment (lululxvi) [inline deepxde/model.py:534]: Combine this with tensorflow above

Comment (HydrogenSulfate) [reply deepxde/model.py]: Done

File: deepxde/model.py (hunk 27 above)
Comment (HydrogenSulfate) [reply deepxde/model.py]: Done
